src/Statement/Grouper/FilteringStatementGrouper.php

```diff
@@ -0,0 +1,101 @@
+<?php
+
+namespace Wikibase\DataModel\Services\Statement\Grouper;
+
+use InvalidArgumentException;
+use Wikibase\DataModel\Services\Statement\Filter\StatementFilter;
+use Wikibase\DataModel\Statement\Statement;
+use Wikibase\DataModel\Statement\StatementList;
+
+/**
+ * @since 3.2
+ *
+ * @licence GNU GPL v2+
+ * @author Thiemo Mättig
+ */
+class FilteringStatementGrouper implements StatementGrouper {
+
+	/**
+	 * @var StatementList[]
+	 */
+	private $groups = array();
+
+	/**
+	 * @var StatementFilter[] An associative array, mapping statement group identifiers to filters.
+	 */
+	private $filters = array();
+
+	/**
+	 * @var string
+	 */
+	private $defaultGroupIdentifier = 'statements';
+
+	/**
+	 * @see \Wikibase\DataModel\Services\Statement\Filter\StatementFilter
+	 *
+	 * @param array $filters An associative array, mapping statement group identifiers to either
+	 *  StatementFilter objects, or to null for the default group.
+	 *
+	 * @throws InvalidArgumentException
+	 */
+	public function __construct( array $filters ) {
+		$this->setFilters( $filters );
+	}
+
+	/**
+	 * @param array $filters
```

Comment:
Unfortunately PHP tools do not understand that. What is currently done is the best you can do.

And yeah, that's really sad. As is the rest of PHPs totally broken collection support :(